docs/pallas/gpu/reference.md

```diff
@@ -162,6 +307,18 @@ There are three operations that can complete a barrier:
   Failing to satisfy this will corrupt the data structure and can cause surprising failures
   (including CUDA runtime errors). See below for an example of a valid program with two threads.
 
+> Another critical restriction is that the number of barrier completions must equal the
+  number of barrier waits throughout its lifetime. It is not allowed to end a scoped
+  allocation of a barrier when it has an unawaited completion. It can be subsequently
+  reused by the compiler and leaving it in this state can cause downstream problems.
+
+> Finally, it is crucial to ensure that each thread that ever waits on a `Barrier`
+  takes part in all `wait` operations on it. It is not allowed to e.g. await every
+  other completion of a barrier from one thread, and all other completions from another
```

Comment:
Curious whether we define the term "thread" anywhere in this documentation? I'm guessing that the reader will be familiar with "CUDA threads" and associate it with lanes, which may make this a little confusing given that the user semantics are not on lanes.

We should make sure to clarify what we mean by the term "thread" somewhere if not already done :)

Reply:
We do, but I added a note to emphasize this even stronger.